### DISCUSSION TITLE
TM4C129 family support

### DISCUSSION THREAD

**Discussion by scaprile** [hathach/tinyusb#1813] [Q&A] [answered]
(Already browsed #670)
What is the support status of TM4C129 family ? I'm particularly interested in TM4C1294 (NDA required)

https://github.com/hathach/tinyusb/blob/48f4d8b7f7eda7ee4046c1c6cd9be2b5c0d81603/src/portable/mentor/musb/musb_tm4c.h#L34-L38

I uncommented the elif and included tm4c129.h and my project builds but nothing happens on the USB side.

Count on me for testing and whatever is needed.

**Comment by hathach** [comment]
I forgot the detail, but if I recall correctly, tm4c 129x has the same usb controller as 123x. Unfortunately I don't have any 129x boards therefore it is not **tested**. Though with a correct  clock/peripheral setup and an a bit troubleshooting maybe you could pull it off.

**Reply by scaprile** [reply]
Thanks,

The 123x has a USB PLL, the 1294 does not, from the datasheet:
> The USB Clock Control Register (USBCC) register contains a CLKDIV bit field which can be programmed to specify the divisor used to reduce the PLL VCO output to the 60 MHz clock source required for the serialization/deserialization module of the USB controller.

That register is part of the USB controller and therefore not documented in the freely available datasheet. I debugged TI code and managed to set that register properly.

Then, TinyUSB asserts in `process_ep0()`,  here:
https://github.com/hathach/tinyusb/blob/301b6af6c6eab582ab027d5cc421eb9215833caf/src/portable/mentor/musb/dcd_musb.c#L486
with req = 255

This controller is a bit different from the one in the 123, it supports HS and the DMA structure is different. Attached is a diff of the USB section in the CMSIS include files: [diff123g1294n.zip](https://github.com/hathach/tinyusb/files/10304454/diff123g1294n.zip)

Had to also fix a mistake in the CMSIS include file, the offset for register CC was incorrect:
```text
1141c1141
<   __I  uint16_t  RESERVED59[77];
---
>   __I  uint16_t  RESERVED59[75];
```

I'm porting working RNDIS examples of [Mongoose](http://mongoose.ws) based on its own TCP/IP stack. I'm also willing to invest time on making this work, though I'm null at USB controllers and reluctant to sign an NDA with TI. Can you provide further help ?
Regards

**Reply by hathach** [reply]
> Then, TinyUSB asserts in process_ep0(), here:
> with req = 255

it asserts due to incorrect received bytes count and req=255 is also incorrect data. This never exist in an actual usb transfer, you probably didn't initialize/configure controller/clock correctly and get garbage data in the register.

**Reply by hathach** [reply]
PS: I didn't know we need to sign NDA to get access to the full manual. This probably makes it harder to support this mcu

**Reply by scaprile** [reply]
The datasheet states the NDA is required. @pran005 seems to have signed it: https://github.com/hathach/tinyusb/pull/670#issuecomment-783249068, though I don't know if he added something later on his own.

My clock settings and USB0->CC settings are the same as those from a working TI sample, and I didn't see further hidden setups, will try step-debugging again

**Reply by kkitayam** [reply]
> with req = 255

255 is REQUEST_TYPE_INVALID.

The IRQ handler or TinyUSB stack may have been invoked with a delay. Please see also https://github.com/hathach/tinyusb/pull/1220#issuecomment-990934381
MUSB handles a part of control transfer automatically. So, when the IRQ handler call is delayed, there will be a state mismatch between the TinyUSB stack and the MUSB IP. If it fails to recover from the state mismatch, host will issue bus reset and retry enumeration process.

It may be improved by setting the USB IRQ priority higher.

**Reply by scaprile** [reply]
Thanks @kkitayam , thank you very much for dropping by.
I have only one interrupt and it is the USB one. I followed the TI sample down to where it starts enabling the bus as TinyUSB does, replicated that (nothing out of the ordinary besides some nonexistent register settings, their library seems to handle 123 and 129 without much questioning except for DMA registers where they ask for controller version and this one is the second generation). I keep getting the same results (or the lack of them).
I've sniffed the bus with Wireshark, I can't see the bus itself. I see there is a device detection and then there is no response to the first descriptor request, probably because of the assertion.
I'm calling tusb_task() in a tight loop after Mongoose event manager, and my USB IRQ handler calls TinyUSB handler; this very same scheme works OK with the STM32F4/7 and the RP2040, though in the RP case their SDK sets the IRQ stuff for me. As I said, I have only one interrupt active and it is the USB one (besides the omnipresent SysTick, of course), but nevertheless I'll follow your advice and raise its priority for the sake of trying. If there is some log I can generate or signals to check, just let me know, I have a modest logic analyzer and a nice oscilloscope, I just don't know what to check. (no USB analyzer, sorry, my logic analyzer can't decode USB).

**Reply by scaprile** [reply]
@kkitayam , after checking, I'm already running USB0 with highest priority as it is the default:
USB priority: 0, SysTick priority: 7

https://e2e.ti.com/support/microcontrollers/arm-based-microcontrollers-group/arm-based-microcontrollers/f/arm-based-microcontrollers-forum/1183688/tm4c1294ncpdt-tinyusb

**Reply by martisen** [reply]
Hi,

I've built tinyUSB for ek-tm4c1294xl (without access to the NDA datasheet) which seems to be working as it should for the example projects.

However, I'm wondering if someone knows where the CMSIS header files have been obtained from? I cannot find them on TIs pages, and in their forum they seem to claim that they don't support this type of header files.
https://github.com/hathach/ti_driver

I have found
https://github.com/speters/CMSIS/tree/master/Device/TI/TM4C/Include
But in these files, there is an error in the USB0_Type struct, causing a lot of the USB registers to be offset.

**Reply by scaprile** [reply]
Oh... I do have the NDA datasheet, and I didn't notice that error (and I'm obsessive...), thank you. I just checked access to the main init registers was to the correct addresses.
I've been using those very same files for Cesanta's Mongoose examples. I remember running a comparison against Keil pack files and only seeing a difference in the GPIO definition (they solve what initially prevents using bit access in writing). I can elaborate offline if you like.

AFAIK there is no TI official CMSIS for that chip, they insist in we using their TIVAware with their names and associated stuff; they even told me (in a very polite way, of course) they don't support register access to their (these ?) hardware products, only through their own firmware.

Can you share the "fixes" ?

**Reply by martisen** [reply, marked as answer]
I havent' found any other errors in the files, but the usage have been limited to SYSCTL and GPIO apart from the USB.
The error is on line 844:
__O  uint8_t   CSRH0;                             /*!< USB Control and Status Endpoint 0 High                                */
  __I  uint16_t  RESERVED24[3];
  __IO uint8_t   COUNT0;                            /*!< USB Receive Byte Count Endpoint 0                                     */
  RESERVED24[3] must be changed to   RESERVED24[2]

**Reply by scaprile** [reply]
Thanks! That "little mistake" translates to all registers that follow... that is why mechanics take all measurements to the same point...
Will get back once I've tried

**Reply by scaprile** [reply]
YES !
Up and running, I sort of corrected for CC offset in this comment above ( https://github.com/hathach/tinyusb/discussions/1813#discussioncomment-4498948 ) but somehow missed which register to correct...
Checked most registers against the datasheets and offsets are correct.
Thank you again, @martisen
